keps/sig-node/4569-cgroup-v1-deprecation/README.md

```diff
@@ -0,0 +1,685 @@
+      - [Integration tests](#integration-tests)
+      - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Deprecate the support for cgroup v1 in Kubernetes, aligning with the industry's move towards cgroup v2 as the default for Linux kernel resource management and isolation.
```

Comment:
Citation needed? AFAICT most projects still support cgroup v1, hybrid, cgroup v2 and we'd be the first mover on dropping it.

This also defacto raises the minimum kernel version quite a bit AFAIK, our docs say cgroup v2 is 5.8+

Right now we require something much much lower: https://github.com/kubernetes-csi/driver-registrar/blob/795af1899f3c94dd0c6dda2a25ed301123479bb9/vendor/k8s.io/kubernetes/test/e2e_node/system/types_unix.go#L33

Reply:
https://github.com/systemd/systemd/issues/30852 might be a good citation

Reply:
The thing that's hard is turning off the cgroup v1 support _by default_. However, we could do things like:
- add a feature gate, on by default, for cgroup v1 support
- detect when Kubernetes (ie the kubelet) is actually on a Linux node using that cgroup v1 support
  - allow people to force the kubelet to fail to start in that case (by disabling the feature gate)
  - if they leave the default in place, complain (fire an Event at kubelet start, annotate the Node, whatever)
    (but keep the kubelet running)
  - maybe even emit a `Warning:` header on every **get** (or other operation) of a node in the deprecated cgroup v1 mode.
- work with SIG Cluster Lifecycle so that tools like `kubeadm` set the feature gate _off_ by default, and need an extra command line argument to permit turning this feature gate on

Reply:
RHEL 9.4 has deprecated cgroup v1 support.

> cgroupsv1 is now deprecated in RHEL 9

> The cgroups is a kernel subsystem used for process tracking, system resource allocation and partitioning. Systemd service manager supports booting in the cgroups v1 mode as well as in cgroups v2 mode. In Red Hat Enterprise Linux 9, the default mode is v2. In Red Hat Enterprise Linux 10, systemd will not support booting in the cgroups v1 mode and only cgroups v2 mode will be available.

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9-beta/html/9.4_release_notes/deprecated-functionality

I am also working on bringing this up to Fedora to see if we can get the distributions to start formalize this. https://discussion.fedoraproject.org/t/is-it-time-to-deprecate-cgroup-v1-support/112875/3

Reply:
Has there been any update on when systemd plans to drop the support of cgroup v1? I imagine that would be key for various OS distros in deciding when to drop as well.

Reply:
https://github.com/systemd/systemd/pull/31442

This went into main a few months ago in systemd. Change was to not boot by default on cgroupv1.

Reply:
Yeah, I was thinking about the actual EOL of cgroup v1 support in systemd. I know there's nothing official announced yet, but wondering if anyone who's active in the systemd community can share some tentative plans/timelines :)